readme.txt
Commands

  Cat Game crafting
    $add floor [min] [max] - add decos to list
      if neither min/max - add whole floor
      if min given but no max - add one deco
      if min and max given - add range of decos
    $remove floor [min] [max] - remove decos
    $clear - clear all decos from list
    $show - show decos in list and crafts needed
    $recipe [floor] [deco] - get one recipe
    optional flags for all commands: 
      -u user - update list for user other than msg author

  Dog Game crafting - same as CG but with word "Dog"
    $dogadd floor [min] [max]
    $dogremove floor [min] [max]
    $dogclear
    $dogshow
    $dogrecipe

  CE crafting
    $starsmax tokens gems hours multiplier [-flags]
      optional flags:
        -a or -aggressive - start crafts early, 1 extra cycle
        -r or -room - add crafting CE room to plan
    $starsgoal goal tokens hours multiplier [-flags]
      optional flags:
        -a or -aggressive - start crafts early, 1 extra cycle  

  Hoard tracking (for TCC)
    $hoard - add lines
    $hoardremove - subtract lines
    $hoardshow - show total and your own lines
    $hoardall - show everyone's hoard contributions (secret command)
    $hoardclear - clear all hoard lines, reset count to 0 (secret command)

  CE crafting (for TCC)
    $room
    $stars3k
  
  TCC admin
    $multiplier - set crafting multiplier
    $pot - add mats to community pot
    $potshow
    $potclear

  Help Commands
    $help - show main help text
    $helpstars - info on $starsmax and $starsgoal
    $helptcc - show TCC CE crafting help
    $helpbadge - for badge admins, list badge categories and nums
    $dm - bot opens a dm
    $calendar - prints event/day for CG and DG
    $hello - check if bot is online

  Badges
    Badge categories (for users to see their badges)
      $easter
      $catlorette
      $mybadges
      $holiday
      $fitness
    Admin
      $badgeadd @user category badgenum - supports multi (comma delim, no spaces)
      $badgeclear @user category
      

DB structure 
(userIDorN specifies it might be a one-word string name if using -u flag)
  crafting~[userIDorN]~cat = list of Cat Game decos
  crafting~[userIDorN]~dog = list of Dog Game decos
  hoard = hoard total
  hoard~[userIDorN] = the user's hoard total
  hoarddisp~[userIDorN] = the user's display name 
  badge~[userID]~category = list of badges
  multiplier = TCC crafting multiplier
  community pot = mats in community pot

ID dictionary
  Channels:
    870785964413452318 - Fearless Revival crafting-bot
    874430786428882944 - Fearless Retrievers crafting-bot
    814911496231845904 - TCC Sweetling's CE bot
    814346827938660354 - Sweetling Sandserver general
    783900285436559360 - Orange's testing channel
  Users:
    721843920404742205 - Sweetling
    179430266895466506 - Orange
    655087300140597268 - Akat


changelog.txt

2022/08/17 split long craft lists into multiple messages, new random quote
2022/07/20 badges - 2022 catlorette rings
2022/07/17 more 2022 catlorette badges
2022/07/16 added 2022 catlorette badges, moved catlorette badges to folder
2022/07/14 bunch of updates
  cat paid floors added: archery, balconygarden, catcafe, corn, duck, outdoorpool, waterfight
  cat prestige added: 
    atlantisprestige, autumnwalkprestige, balconygardenprestige, breadprestige, 
    minobeachprestige, minoforestprestige, outdoorpoolprestige,
    bakingprestige, yardsaleprestige, deliprestige, rainyprestige, minigolfprestige
  dog paid floors added: catfloor, dogspa, donutfloat, highfashion, kpop, stargazing, tarot
  dog tower floors added: sugarland, wrestling, hairsalon, beach
2022/02/15 add hoard commands
2022/02/14 removed banned servers, switched to new bot (new token)
2022/02/13 added starsgoal
2022/02/11 
  bot back up except in fearless whispurrs discord
  orange removed from badge admin
  starsmax allowed in all channels
2022/02/09 brought down bot - display error msg always
2022/01/25 refactored badge code and categories, added fitness category and badge 1 
2022/01/19 dg cosmicunicorns floor
2022/01/12 doggydaycare, supercatsquadprestige, asianfoodprestige
2022/01/01 multi badge add
2021/12/31 FR anniversary and lunar new year badges
2021/12/30 yule and festivus badges
2021/12/29 gold and silver medal badges, las posadas and soyal badges, pink egg
2021/12/28 diwali and eid badges
2021/12/27 holiday badge category; christmas, hannukah, kwanzaa badges; new quote
2021/12/09 added rainy, minigolf, hatcatsprestige floors in cat. added craft total to display message
2021/11/29 added blue frgem2 badge, allow orange and akat to assign badges
2021/11/27 added frgem badge
2021/11/22 
  added dinosaurs, heavymetal, gemstones floors in dog
  added freshfruit floor in cat
  updated helptext for starsmax in tcc
  random quotes for access denied msg
  removed starsmax from FR channels and enabled in DMs
2021/11/05 added minigame participation (phoenix ruff) badge images/phoenixruff.png
2021/11/04 
  deprecated the old starsmax
  added a new starsmax in new file clubevent.py that is way stronger
  updated messages and readme to match
2021/10/18 added aquarium, autumn walk, super cat squad in cat recipes
2021/08/13 added calendar tag for telling what day/event it is
2021/08/11 added deli floor in cg
2021/07/29 
  added dog game support and recipes
  refactored code to take 'cat' or 'dog' as param, for dog game support
  fixed bugs in tattoo shop and airport CG recipes
  fixed bug that never counted bronze (i typed copper in a key spot)
  refactored badge code
  updated help text
2021/07/29 created change log. prior to now we have cat game crafting recipes, TCC crafting schedules for room, 3k, and starsmax, and some badges (easter and catlorette)


test.py
from replit import db
keys = db.keys()

for key in keys:
  break
  if 1==0: #key.startswith("badges~") and ("catlorette" in key):
    #msg = str(key) + "---" + str(db[key]) + "\n"
    #print(msg)
    badgeAry = db[key]
    for i in range(len(badgeAry)):
      pieceAry = badgeAry[i].split("/")
      if len(pieceAry)==2:
        badgeAry[i]='images/catlorette/'+pieceAry[1]
    #print(badgeAry)
    db[key] = badgeAry

for key in keys:
  if key.startswith("badges~") and ("general" in key):
    if "#" in key: continue
    del db[key]
    continue
    #msg = str(key) + "---" + str(db[key]) + "\n"
    #print(msg)
    keypieces = key.split("~")
    newkey = "badges~"+str(keypieces[1])+"~mybadges"
    db[newkey] = db[key]

print("done")

tcc.py
import math

def room(multiplier):
  outmsg="\n**Craft the room in 2 hours**"
  outmsg=outmsg+"\nTimes given as minutes after start of boost" 

  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n0:00 - 75/75/75 ABC"
  outmsg=outmsg+"\n0:15 - 75/75/75 ABC, 20D"
  outmsg=outmsg+"\n0:30 - 75/75/75 ABC, 20D"
  outmsg=outmsg+"\n0:45 - 75/75/75 ABC, 20D, 18E"
  outmsg=outmsg+"\n1:00 - 75/75/75 ABC, 20D"
  outmsg=outmsg+"\n1:15 - 75/75/75 ABC, 20D, 18E"
  outmsg=outmsg+"\n1:30 - 75/75/75 ABC, 20D"
  outmsg=outmsg+"\n1:45 - 20D, 18E, 11F"
  
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\nTotal crafted - 525/525/525 ABC, 140D, 54E, 11F"
  outmsg=outmsg+"\nInventory - 17/29/29 ABC, 32D, 21E, 11F"
  outmsg=outmsg+"\nMultiplier: " + str(multiplier)
  outmsg=outmsg+"\nCost: " + str(getRoomCost(multiplier))

  return outmsg


def stars3k (multiplier):
  outmsg="\n**Craft 3000 stars (1000E) in 5 hours**"
  outmsg=outmsg+"\nTimes given as minutes after start of boost" 

  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n0:00 - 392/504/504 ABC"
  outmsg=outmsg+"\n0:15 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n0:30 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n0:45 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n1:00 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n1:15 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n1:30 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n1:45 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n2:00 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n2:15 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n2:30 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n2:45 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n3:00 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n3:15 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n3:30 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n3:45 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n4:00 - 392/504/504 ABC, 112D"
  outmsg=outmsg+"\n4:15 - 392/504/504 ABC, 112D, 112E"
  outmsg=outmsg+"\n4:30 - 10/10/10 ABC, 112D"
  outmsg=outmsg+"\n4:45 - 112E"
  
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\nTotal crafted - 7066/9082/9082 ABC, 2016D, 1008E"
  outmsg=outmsg+"\nInventory - 1018/1018/1018 ABC, 1008E"
  outmsg=outmsg+"\nMultiplier: " + str(multiplier)
  outmsg=outmsg+"\nCost: " + f'{get3kCost(multiplier):,}'

  return outmsg

def getRoomCost(multiplier):
  base = 807100
  recipes = 6430
  return int((base/multiplier) + recipes)

def get3kCost(multiplier):
  base = 74420640 + 1950 #1950 is the 3 stacks of 10 for ABC buffer
  starRecipe = 300*1008
  return int((base/multiplier) + starRecipe)

def baseCost(craft):
  switcher = {
    "A": 10,
    "B": 10,
    "C": 10,
    "D": 75,
    "E": 100,
    "F": 200
  }
  return switcher.get(craft,0)
  
def stackCost(craft, stack):
  base=baseCost(craft)
  return int((stack+3)*base*stack/2)

def getStacks(stackEper30min):
  stackA=math.ceil(7*stackEper30min/2)
  stackB=math.ceil(9*stackEper30min/2)
  stackC=math.ceil(9*stackEper30min/2)
  return [stackA,stackB,stackC]

def costPer15min(stackEper30min, multiplier):
  # base crafting cost per 30 min
  aryABC=getStacks(stackEper30min)
  cost=2*stackCost("A",aryABC[0])
  cost=cost+2*stackCost("B",aryABC[1])
  cost=cost+2*stackCost("C",aryABC[2])
  cost=cost+2*stackCost("D",stackEper30min)
  cost=cost+stackCost("E",stackEper30min)

  # adjust for multiplier and stars recipe
  cost = int(cost/multiplier)+(300*stackEper30min)
  return int(cost/2)

def findEper30min(tokensPerRound,multiplier):
  mult20exceed = 20

  while costPer15min(mult20exceed,multiplier)<tokensPerRound:
    mult20exceed=mult20exceed+20
  stackEper30min = mult20exceed-20

  mult5exceed=stackEper30min
  while costPer15min(mult5exceed,multiplier)<tokensPerRound:
    mult5exceed=mult5exceed+5
  stackEper30min = mult5exceed-5

  return stackEper30min

def starsmax(args,multiplier):
  argStr=args.strip()
  if len(argStr)==0: return "Expected syntax: $starsmax [tokens] or $starsmax [tokens] [multiplier]"

  argAry=argStr.split(" ")
  try: tokens=int(argAry[0].replace(',', ''))
  except: return errtcc()
  
  if len(argAry)>1: mult=float(argAry[1])
  else: mult=multiplier
  outmsg="\n**Maximize token use in 5 hours**"
  outmsg=outmsg+"\nTimes given as minutes after start of boost" 

  tokensPerRound=int(tokens/18)
  stackEper30min=findEper30min(tokensPerRound,mult)
  aryABC=getStacks(stackEper30min)
  stackA=aryABC[0];stackB=aryABC[1];stackC=aryABC[2]
  stackD=stackEper30min

  strABC=str(stackA)+"/"+str(stackB)+"/"+str(stackC)+" ABC"
  strD=str(stackD)+"D";strE=str(stackEper30min)+"E"

  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n0:00 - " + strABC
  outmsg=outmsg+"\n0:15 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n0:30 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n0:45 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n1:00 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n1:15 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n1:30 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n1:45 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n2:00 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n2:15 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n2:30 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n2:45 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n3:00 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n3:15 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n3:30 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n3:45 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\n4:00 - " + strABC + ", " + strD 
  outmsg=outmsg+"\n4:15 - " + strABC + ", " + strD + ", " + strE 
  outmsg=outmsg+"\n4:30 - 10/10/10 ABC, " + strD
  outmsg=outmsg+"\n4:45 - " + strE
  
  totalA=18*stackA+10;totalB=18*stackB+10;totalC=18*stackC+10
  totalD=18*stackD;totalE=9*stackEper30min
  strABC=str(totalA)+"/"+str(totalB)+"/"+str(totalC)+" ABC"
  strD=str(totalD)+"D";strE=str(totalE)+"E"
  outmsg=outmsg+"\n"
  outmsg=outmsg+"\nTotal crafted - " + strABC + ", " + strD + ", " + strE

  invA=totalA-(6*totalE);invB=totalB-(8*totalE);invC=totalC-(8*totalE)
  invD=totalD-(2*totalE);invE=totalE
  strABC=str(invA)+"/"+str(invB)+"/"+str(invC)+" ABC"
  strD=str(invD)+"D";strE=str(invE)+"E"
  outmsg=outmsg+"\nInventory - " + strABC + ", " + strD + ", " + strE
  outmsg=outmsg+"\nMultiplier: " + str(mult)

  totalCost = 18*costPer15min(stackEper30min,mult)
  extraCostBase = stackCost("A",10) + stackCost("B",10) + stackCost("C",10)
  totalCost = totalCost + int(extraCostBase/mult)
  outmsg=outmsg+"\nCost: " + f'{totalCost:,}'

  return outmsg

def helptcc():
  outmsg="\n**TCC Commands**\n"
  outmsg=outmsg+"These commands are for use in the TCC server."

  outmsg=outmsg+"\n\n"
  outmsg=outmsg+"**"+"Commands"+"**"
  outmsg=outmsg+"\n"

  outmsg=outmsg+"**$room**: Step-by-step instructions to craft the room in 2 hours.\n"
  outmsg=outmsg+"**$stars3k**: Step-by-step instructions to craft 3000 stars (1000E) in 5 hours.\n"
  outmsg=outmsg+"**$starsmax**: Step-by-step instructions to craft as many E as you have tokens and gems for. The cost of the star recipe is included in these calculations. Enter $helpstars for more info.\n"

  outmsg=outmsg+"\n\n"
  outmsg=outmsg+"**"+"Admin Commands"+"**"
  outmsg=outmsg+"\n"

  outmsg=outmsg+"**$multiplier**: Enter *$multiplier [x]* to set the multiplier for the server, for example *$multiplier 30.1*.\n"
  outmsg=outmsg+"**$pot**: Enter *$pot [x]* to add x mats to the community pot, for example *$pot 100*.\n"
  outmsg=outmsg+"**$potshow**: Show what's in the community pot.\n"
  outmsg=outmsg+"**$potclear**: Clear the community pot.\n"

  return outmsg

def errtcc():
  return "Something went wrong. Enter *$helptcc* for instructions."

main.py
import discord
import os
import redis

from os import environ
#environ["REPLIT_DB_URL"] = "https://kv.replit.com/v0/eyJhbGciOiJIUzUxMiIsImlzcyI6ImNvbm1hbiIsImtpZCI6InByb2Q6MSIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJjb25tYW4iLCJleHAiOjE2NjMxOTk1MzAsImlhdCI6MTY2MzA4NzkzMCwiZGF0YWJhc2VfaWQiOiI5NWI0MjhiNy04NGE3LTQ5ODktYmZiZC01NDNhMzcxYWQzNmIiLCJ1c2VyIjoiU2NvdHROaWVtYW5uIiwic2x1ZyI6IlN3ZWV0Qm90In0.PqKEYwiSkGuYswO7NhMV5utCgk8PAhNoV81OcJCIv6FaZe7emOOG9M1KutF-n9GV_nvRLO6ordKFMq9sQdm37A"
#environ["REPLIT_DB_URL"] = "https://kv.replit.com/v0/eyJhbGciOiJIUzUxMiIsImlzcyI6ImNvbm1hbiIsImtpZCI6InByb2Q6MSIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJjb25tYW4iLCJleHAiOjE2NjMzMjIyNzIsImlhdCI6MTY2MzIxMDY3MiwiZGF0YWJhc2VfaWQiOiI5NWI0MjhiNy04NGE3LTQ5ODktYmZiZC01NDNhMzcxYWQzNmIiLCJ1c2VyIjoiU2NvdHROaWVtYW5uIiwic2x1ZyI6IlN3ZWV0Qm90In0.FSoG-qK70XAq837M-rHcL63XxGi5dVGpngvjsYbTBQOHBLkUQePCcDdkCByRQpz8zxSsM80gsk_79VyXwwOt-g"
#from replit import db
from messages import mainHelpText, calendar, starsmaxDetail, randQuote, helpbadge
from recipes import getCraftsCat, getRecipeCat, floorSpellcheckCat
from recipesdog import getCraftsDog, getRecipeDog, floorSpellcheckDog
from tcc import room, stars3k, helptcc, errtcc
from clubevent import maxstarshelper,goalstarshelper
from badges import badgeAddMulti, badgeClear, getBadges
from keep_alive import keep_alive
r = redis.from_url(os.environ.get("REDIS_URL"))

def Convert(string):
    li = list(string.split(" "))
    return li
  


def add_decos(user,decos,category):
  if category=="": category='cat'
  argAry=interpretArgs(decos,category)
  if len(argAry)==0: return "Something went wrong. Expected syntax: $add floorname [startingdeco] [endingdeco]\nEnter floor name as one word"

  floor=argAry[0]
  if getRecipe(category,floor,1)=="Invalid floor": return "Floor name not recognized. Check spelling, and enter the floor name as one word."
  minDeco=int(argAry[1])
  maxDeco=int(argAry[2])

  alreadypresent=""
  userkey=getUserKey(user,category)
  if r.exists(userkey):
    decoslist = Convert(r.get(userkey).decode())
  else:
    decoslist = []
  
  for x in range (minDeco, maxDeco+1):
    hash=floor+"/"+str(x)
    if hash not in decoslist:
      decoslist.append(hash)
    else:
      alreadypresent=alreadypresent+", "+hash
  #decoslist.sort()
  #it was required to move to sorted per heroku recommendation
  #sorted(decoslist);
  decoslist.sort()
  r.set(userkey, ' '.join(decoslist))
  #print(r.get(userkey))

  if len(alreadypresent)==0:
    return "Decos added. Happy crafting!"
  return "Decos added. Happy crafting! Decos already in your list:"+alreadypresent.split(",",1)[1]

def remove_decos(user,decos,category):
  if category=="": category='cat'
  userkey=getUserKey(user,category)
  if not r.exists(userkey): return "You do not have any decos in your list."
 
  decoslist = Convert(r.get(userkey).decode())
  #decoslist = db[userkey]
  
  argAry=interpretArgs(decos,category)
  if len(argAry)==0: return "Expected syntax: $remove floorname [startingdeco] [endingdeco]"
  floor=argAry[0]
  if getRecipe(category,floor,1)=="Invalid floor": return "Floor name not recognized. Make sure to enter the floor name as one word."
  minDeco=int(argAry[1])
  maxDeco=int(argAry[2])

  for x in range (minDeco, maxDeco+1):
    hash=floor+"/"+str(x)
    if hash in decoslist:
      decoslist.remove(hash)
  
  #db[userkey] = decoslist
  r.set(userkey, ' '.join(decoslist))
  return "Decos removed. Yeet!"

def clear_decos(user, category):
  if category=="": category='cat'
  userkey=getUserKey(user, category)
  if r.exists(userkey): r.delete(userkey)
  #if userkey in db.keys(): del db[userkey]
  return "Decos cleared."

def show_msg(user,category):
  if category=="": category='cat'
  userkey=getUserKey(user, category)
  if not r.exists(userkey): return ["Planned decos: none"]
  
  decoslist = Convert(r.get(userkey).decode())
  total=len(decoslist)
  if total==0: return ["Planned decos: none"]
    
  msg1 = "Planned decos: "+str(total)+" total"
  msg3 = getCrafts(decoslist,category)

  if total<101:
    crafts = ""
    for x in decoslist: crafts=crafts+", "+str(x)
    crafts = crafts[2:]
    return [msg1 + "\n" + crafts + "\n\n" + msg3]

  else: # split of message as main message first, then list of crafts
    crafts = ""
    returnary=[msg1 + "\n\n" + msg3]
    for x in decoslist: 
      crafts=crafts+", "+str(x)
      if len(crafts)>1970: 
        returnary.append(crafts[2:])
        crafts = ""
    if len(crafts)>0: returnary.append(crafts[2:])
    return returnary
    

def getCrafts(decoslist,category):
  if category=='dog': return getCraftsDog(decoslist)
  return getCraftsCat(decoslist)

def printRecipe(decoStr,category):
  decoAry=decoStr.strip().split()
  if len(decoAry)<2: outmsg="Enter a floor and deco number."
  outmsg=getRecipe(category,decoAry[0], decoAry[1])
  if outmsg=="": outmsg="No recipe found."
  return outmsg

def getRecipe(category,floor,deco):
  if category=='dog': return getRecipeDog(floor,deco)
  return getRecipeCat(floor,deco)

def interpretArgs(decos,category):
  if category=="": category='cat'
  decosAry = decos.split()
  floor=floorSpellcheck(decosAry[0].lower(),category)
  length = len(decosAry)
  if length==1:
    minDeco=1
    maxDeco=7
  elif length==2:   #checks that the deco num is between 1 and 7. Does NOT take a guess.
    try: minDeco=int(decosAry[1])
    except: return []
    if minDeco<1: return []
    if minDeco>7: return []
    maxDeco=minDeco
  elif length==3:   #checks that range is between 1 and 7. Does take guesses.
    try: minDeco=max(int(decosAry[1]),1)   # if min is -5 or something, changes to 1
    except: return []
    try: maxDeco=min(int(decosAry[2]),7)   # if max is 99 or something, changes to 7
    except: return []
  else:
    return []
  return [floor,str(minDeco),str(maxDeco)]

def floorSpellcheck(floor,category):
  if category=='dog': return floorSpellcheckDog(floor)
  return floorSpellcheckCat(floor)

def parseFlags(flagStr):
  tokenAry = flagStr.split()
  tokenCount = len(tokenAry)
  if tokenCount==0: return [[],[]]
  flagAry = ["" for i in range(tokenCount)]
  valAry = ["" for i in range(tokenCount)]
  for i in range(tokenCount):
    token = getAlias(tokenAry[i].strip().lower())
    if token=="-user": #deco craftin - flag is -user and next word is username
      if i+1 < tokenCount: 
        flagAry[i] = token
        valAry[i] = tokenAry[i+1].strip()
    if token in ["-room","-aggressive","-debug"]: #CE options
      flagAry[i] = token
  return [flagAry,valAry]

def getUserKey(user,category):
  if category=="": category='cat'
  return 'crafting~'+str(user)+"~"+category

def getAlias(flag):
  switcher = {
    "-u": "-user",
    "-a": "-aggressive", "-ag": "-aggressive", "-aggr": "-aggressive",
    "-r": "-room", "-rm": "-room"
  }
  return switcher.get(flag,flag) #swaps the flag if it's found, otherwise unchanged

def showBadges(user,category):
  badgeary=getBadges(user,category)
  if badgeary[0]!="": return ["No "+badgeary[0]+" found. Keep collecting!", "", ""]

  embedmsg = discord.Embed(
    title=badgeary[2])
  #file = discord.File("images/egg1.png", filename="image.png")
  file = discord.File(badgeary[1], filename="image.png")
  embedmsg.set_image(url="attachment://image.png")
  return ["", file, embedmsg]

# it was necessary to delcare message intent (appropriate discord bot also need to be set in developer portal
intents = discord.Intents.default()
intents.message_content = True
client = discord.Client(intents=intents)

@client.event
async def on_ready():
  print('We have logged in as {0.user}'.format(client))

@client.event
async def on_message(message):

  if message.author == client.user: 
    return
  
  msg = message.content
  print(message.author)
  print(message.content)

  
  if not msg.startswith('$'):
    return

    
  if msg.startswith('$hello'):
    await message.channel.send('Hello, world!')
    return

  if msg.startswith('$dm'):
    channel = await message.author.create_dm()
    await channel.send('Welcome to Crafting Bot! Enter *$help* for instructions!')
    
  # $help: prints instructions
  msg = message.content
  if msg.startswith('$help'):
    if msg.startswith('$helptcc'): 
      await message.channel.send(helptcc()) 
      return
    if msg.startswith('$helpstar'): 
      ary=starsmaxDetail()
      await message.channel.send(ary[0]) 
      await message.channel.send(ary[1]) 
      return
    if msg.startswith('$helpbadge'): 
      await message.channel.send(helpbadge()) 
      return

    ary=mainHelpText()
    await message.channel.send(ary[0]) 
    await message.channel.send(ary[1]) 
    return

  if msg.startswith('$calendar'):
    await message.channel.send(calendar()) 
    return

  #tcc commands - not user specific
  if msg.startswith('$multiplier'):
    try: mult = (float)(msg.split("$multiplier",1)[1])
    except: 
      await message.channel.send(errtcc())
    #db["multiplier"] = mult
    r.set("multiplier", ' '.join(mult))
    await message.channel.send("Multiplier set")
    return

  if msg.startswith('$room'):
    await message.channel.send(room(r.get("multiplier").decode()))
    return

  if msg.startswith('$stars3k'):
    await message.channel.send(stars3k(r.get("multiplier").decode()))
    return

  if msg.startswith('$pot'):
    if msg.startswith('$potclear'):
      db["community pot"] = 0
      r.set("community pot", 0)
      await message.channel.send("Community pot cleared.")
      return
    if msg.startswith('$potshow'):
      await message.channel.send("Community pot: "+ str(r.get("community pot").decode()))
      return
    mats = msg.split("$pot",1)[1]
    #db["community pot"] = int(db["community pot"])+int(mats)
    await message.channel.send("Mats added to community pot.")
    return

  #
  user=str(message.author.id)
  userflag=0
  flagAry = []
  if "-" in msg:
    msgAry = msg.split("-",1)
    msg=msgAry[0].strip()
    flagStr="-"+msgAry[1]
    parseAry=parseFlags(flagStr)
    flagAry=parseAry[0]
    valAry=parseAry[1]
    
    try: 
      userIndex=flagAry.index("-user")
      user=valAry[userIndex].lower()
      userflag=1
    except: userIndex=-1 

  # $starsmax tokens gems hours multiplier [-room] [-aggressive]
  if msg.startswith('$starsmax'):
    if (1==1): #now allowed in all channels
      args = msg.split("$starsmax",1)[1]
      await message.channel.send(maxstarshelper(args,flagAry))
      return

    #the original starsmax in tcc.py was deprecated
    #await message.channel.send(starsmax(args,db["multiplier"]))
    await message.channel.send(randQuote()) # if it's in a non-allowed channel
    return

  # $starsgoal goal tokens hours multiplier [-aggressive]
  if msg.startswith('$starsgoal'):
    args = msg.split("$starsgoal",1)[1]
    await message.channel.send(goalstarshelper(args,flagAry))
    return

  # $add: add decos to cat game crafting list
  # syntax:
  #   $add floorname      - adds whole floor e.g. $add waterpark
  #   $add floorname deco - adds one deco e.g. $add waterpark 1
  #   $add floorname start end - adds the range of decos from that floor
  # notes: floorname must be all one word e.g. hairsalon not hair salon
  if msg.startswith('$add'):
    decos = msg.split("$add",1)[1]
    await message.channel.send(add_decos(user,decos,'cat'))
    return
  
  # $remove: removes one deco from your cat game list
  # syntax: $remove floorname deco - e.g. $remove waterpark 5
  if msg.startswith('$remove'):
    decos = msg.split("$remove",1)[1]
    await message.channel.send(remove_decos(user,decos,'cat'))
    return

  # $clear: remove all decos from your cat game list
  if msg.startswith('$clear'):
    await message.channel.send(clear_decos(user,'cat'))
    return

  # $show: shows the decos on your cat game list, then a list of crafts required in inventory
  if msg.startswith('$show'):
    returnary = show_msg(user,'cat')
    for msg in returnary: await message.channel.send(msg)
    return
  
  # $recipe: recipe of one specific deco in cat game
  if msg.startswith('$recipe'):
    decoStr = msg.split("$recipe",1)[1]
    await message.channel.send(printRecipe(decoStr,'cat'))
    return

  #the equivalents for dog game
  if msg.startswith('$dogadd'):
    decos = msg.split("$dogadd",1)[1]
    await message.channel.send(add_decos(user,decos,'dog'))
    return
  if msg.startswith('$dogremove'):
    decos = msg.split("$dogremove",1)[1]
    await message.channel.send(remove_decos(user,decos,'dog'))
    return
  if msg.startswith('$dogclear'):
    await message.channel.send(clear_decos(user,'dog'))
    return
  if msg.startswith('$dogshow'):
    returnary = show_msg(user,'dog')
    for msg in returnary: await message.channel.send(msg)
    return
  if msg.startswith('$dogrecipe'):
    decoStr = msg.split("$dogrecipe",1)[1]
    await message.channel.send(printRecipe(decoStr,'dog'))
    return
  #end of dog game section
    


  #Sweetling crafting bot badges :)
  if msg.startswith('$badge'):
    if message.author.id not in [721843920404742205,655087300140597268]: #Only some users can add/remove badges
      await message.channel.send(randQuote())
      return
    if msg.startswith('$badgeadd'):
      params=msg.split("$badgeadd",1)[1].strip()
      paramAry=params.split(" ")
      badgeAddMulti(paramAry[0],paramAry[1],paramAry[2])
      await message.channel.send("Badge added.")
      return
    if msg.startswith('$badgeclear'): 
      params=msg.split("$badgeclear",1)[1].strip()
      paramAry=params.split(" ")
      badgeClear(paramAry[0],paramAry[1])
      await message.channel.send("Badges cleared.")
      return
    if msg.startswith('$badgehelp'):
      await message.channel.send(helpbadge())
    return

  keyword=msg.split()[0][1:]
  badgewords=['mybadges','easter','catlorette','holiday','fitness']
  if keyword in badgewords:
    getuser=msg.split(keyword,1)[1].strip()
    if not getuser: getuser=user
    ary=showBadges(getuser,keyword)
    if ary[0]=="":
      await message.channel.send(file=ary[1],embed=ary[2])
      return
    await message.channel.send(ary[0])
    return

# not necessary on heroku, also remove call to server object
#keep_alive()

# DISCORD_TOKEN defined in heroku env_variables 
client.run(os.environ['DISCORD_TOKEN'])

# the code to check for server, for functions that have to be run from a server
#try:      #checking for server ID
#  serverID=message.guild.id
#except:
#  await message.channel.send('Please use from a server, thanks!')
#  return
#user = str(serverID) + "---" + str(message.author.id)
